akka-sample-kafka-to-sharding-scala/processor/src/main/scala/sample/sharding/kafka/UserEvents.scala

```diff
@@ -46,32 +43,42 @@ object UserEvents {
             runningTotal.copy(
               totalPurchases = runningTotal.totalPurchases + 1,
               amountSpent = runningTotal.amountSpent + (quantity * price)))
-        case GetRunningTotal(_, replyTo) =>
+        case GetRunningTotal(id, replyTo) =>
+          ctx.log.info("user {} running total queried", id)
           replyTo ! runningTotal
           Behaviors.same
       }
     }
   }
 
-  /*
-   * The murmur2 message extractor matches kafka's default partitioning when messages
-   * have keys that are strings
+  /**
+   * Asynchronously get the Akka Cluster Sharding [[ShardingMessageExtractor]]. Given a topic we can automatically
+   * retrieve the number of partitions and use the same hashing algorithm used by the Apache Kafka
+   * [[org.apache.kafka.clients.producer.internals.DefaultPartitioner]] (murmur2) with Akka Cluster Sharding.
    */
-  class UserIdMessageExtractor(nrKafkaPartitions: Int)
-      extends Murmur2NoEnvelopeMessageExtractor[Message](nrKafkaPartitions) {
-    override def entityId(message: Message): String = message.userId
+  def messageExtractor(system: ActorSystem[_]): Future[KafkaClusterSharding.KafkaShardingNoEnvelopeExtractor[Message]] = {
+    val processorConfig = ProcessorConfig(system.settings.config.getConfig("kafka-to-sharding-processor"))
+    KafkaClusterSharding.messageExtractorNoEnvelope(
+      system = system.toClassic,
+      timeout = 10.seconds,
+      topic = processorConfig.topics.head,
+      entityIdExtractor = (msg: Message) => msg.userId,
+      settings = ConsumerSettings(system.toClassic, new StringDeserializer, new StringDeserializer)
+        .withBootstrapServers(processorConfig.bootstrapServers)
+    )
   }
 
-  def init(system: ActorSystem[_]): ActorRef[Message] = {
-    val processorConfig = ProcessorConfig(system.settings.config.getConfig("kafka-to-sharding-processor"))
+  def init(system: ActorSystem[_], messageExtractor: ShardingMessageExtractor[Message, Message], groupId: String): ActorRef[Message] = {
+    // create an Akka Cluster Sharding `EntityTypeKey` for `UserEvents` for this Kafka consumer group
+    val typeKey = EntityTypeKey[UserEvents.Message](groupId)
     ClusterSharding(system).init(
-      Entity(TypeKey)(createBehavior = _ => UserEvents())
-        .withAllocationStrategy(new ExternalShardAllocationStrategy(system, TypeKey.name))
-        .withMessageExtractor(new UserIdMessageExtractor(processorConfig.nrPartitions))
+      Entity(typeKey)(createBehavior = _ => UserEvents())
+        // NOTE: why does `ExternalShardAllocationStrategy` not accept the type key type itself?
+        .withAllocationStrategy(new ExternalShardAllocationStrategy(system, typeKey.name))
```

Comment:
i think it is in the untyped module so doesn't have access to ti

Reply:
Ah, got it.